File: AppSets-Android/appsets/src/main/kotlin/xcj/app/appsets/worker/ServerSyncWorker.kt
package xcj.app.appsets.worker

import android.content.Context
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import xcj.app.appsets.db.room.repository.GroupInfoRepository
import xcj.app.appsets.db.room.repository.UserInfoRepository
import xcj.app.appsets.server.repository.UserRepository
import xcj.app.appsets.usecase.RelationsUseCase
import xcj.app.starter.android.util.PurpleLogger
import xcj.app.starter.server.request

class ServerSyncWorker(
    context: Context,
    params: WorkerParameters,
) : CoroutineWorker(context, params) {
    companion object {
        private const val TAG = "ServerSyncWorker"
    }

    override suspend fun doWork(): Result {
        PurpleLogger.current.d(TAG, "doWork, thread:${Thread.currentThread()}")
        val conditions = inputData.getString("conditions")
        PurpleLogger.current.d(TAG, "doWork, conditions:${conditions}")

        if (conditions.isNullOrEmpty()) {
            return Result.failure()
        }
        var hasSingleCondition = false
        if (conditions.contains("friends")) {
            hasSingleCondition = true
            val friends = request {
                UserRepository.getInstance().getFriends()
            }.getOrNull()
            PurpleLogger.current.d(TAG, "doWork, friends:$friends")
            if (!friends.isNullOrEmpty()) {
                PurpleLogger.current.d(TAG, "doWork, add friends to local room db")
                UserInfoRepository.getInstance().addRelatedUserInfoList(friends)
                RelationsUseCase.getInstance()
                    .initRelatedUsersFromServer(friends.map { it.uid }.toSet())
            }
        }

        if (conditions.contains("groups")) {
            hasSingleCondition = true
            val groups = request {
                UserRepository.getInstance().getChatGroups()
            }.getOrNull()
            PurpleLogger.current.d(TAG, "doWork, groups:$groups")
            if (!groups.isNullOrEmpty()) {
                val groupIdMap = mutableMapOf<String, Set<String>?>()
                groups.forEach {
                    groupIdMap[it.groupId] =
                        it.userInfoList?.map { userInfo -> userInfo.uid }?.toSet()
                }
                PurpleLogger.current.d(TAG, "doWork, add groups to local room db")
                GroupInfoRepository.getInstance().addRelatedGroupInfoList(groups)
                RelationsUseCase.getInstance().initRelatedGroupsFromServer(groupIdMap)
            }
        }
        if (hasSingleCondition) {
            return Result.success()
        }
        return Result.failure()
    }
}

File: AppSets-iOS/AppSets/com/app/appsets/worker/ServerSyncWorker.swift
//
//  ServerSyncWorker.swift
//  AppSets
//
//  Created by caiju Xu on 2024/7/12.
//

import Foundation

class ServerSyncWorker : ListenableWorker {
    
    private static let TAG = "ServerSyncWorker"
    
    override func doWork() async -> Result {
        PurpleLogger.current.d(ServerSyncWorker.TAG, "doWork")
        
        guard let conditions = getInputData().getString("conditions") else {
            
            return Result.Failed
        }
        
        PurpleLogger.current.d(ServerSyncWorker.TAG, "doWork, conditions:\(conditions)")
        
        
        var hasSupportConditions = false
        
        if conditions.contains("friends") {
            hasSupportConditions = true
            let friends:[UserInfo]? = await UserRepository().getFriends()
            PurpleLogger.current.d(ServerSyncWorker.TAG, "doWork, friends:\(String(describing: friends))")
           
            await MainActor.run{
                if let userInfoList = friends {
                    var uids: [String] = []
                    userInfoList.forEach{ userInfo in
                        if !String.isNullOrEmpty(userInfo.uid) {
                            uids.append(userInfo.uid)
                        }
                    }
                    UserInfoRepository().addRelatedUserInfoList(userInfoList)
                    RelationsUseCase.Instance.initRelatedUsersFromServer(uids)
                }
            }
        }
        
        
        if conditions.contains("groups") {
            hasSupportConditions = true
            let groups:[GroupInfo]? = await UserRepository().getChatGrups()
            PurpleLogger.current.d(ServerSyncWorker.TAG, "doWork, groups:\(String(describing: groups))")
            await MainActor.run {
                if let groupInfoList = groups {
                    var groupIdMap: [String: Set<String>] = [String: Set<String>] ()
                    groups?.forEach{ groupInfo in
                        var uids: Set<String> = []
                        groupInfo.userInfoList?.forEach({ userInfo in
                            if !String.isNullOrEmpty(userInfo.uid) {
                                uids.insert(userInfo.uid)
                            }
                        })
                        if !uids.isEmpty {
                            groupIdMap[groupInfo.groupId] = uids
                        }
                    }
                    GroupInfoRepository().addRelatedGroupInfoList(groupInfoList)
                    
                    RelationsUseCase.Instance.initRelatedGroupsFromServer(groupIdMap)
                   
                  
                }
            }
        }
        
        if hasSupportConditions {
            return Result.Success
        }
        
        return Result.Failed
        
    }
    
}
